### PR TITLE
Consider missing attribute in context validation

### PR DESCRIPTION
**Pull request by vjik**
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ❌
| Fixed issues  | #556

### CODE REVIEW

File: src/ValidationContext.php [hunk 1]
```diff
@@ -127,11 +132,14 @@ public function validate(mixed $data, callable|iterable|object|string|null $rule
 
         $currentDataSet = $this->dataSet;
         $currentAttribute = $this->attribute;
+        $currentIsDataSetMissing = $this->isDataSetMissing;
```
Comment (samdark) [inline src/ValidationContext.php:135]:
```suggestion
        $isCurrentDataSetMissing = $this->isDataSetMissing;
```

File: src/ValidationContext.php [hunk 2]
```diff
@@ -127,11 +132,14 @@ public function validate(mixed $data, callable|iterable|object|string|null $rule
...
 
+        $this->isDataSetMissing = $this->isAttributeMissing();
         $result = $this->validator->validate($data, $rules, $this);
 
         $this->dataSet = $currentDataSet;
         $this->attribute = $currentAttribute;
+        $this->isDataSetMissing = $currentIsDataSetMissing;
```
Comment (samdark) [inline src/ValidationContext.php:142]:
```suggestion
        $this->isDataSetMissing = $isCurrentDataSetMissing;
```

**Review by samdark — APPROVED** [review]

File: src/ValidationContext.php [hunk 3]
```diff
@@ -262,7 +270,7 @@ public function setParameter(string $name, mixed $value): self
      */
     public function isAttributeMissing(): bool
     {
-        return $this->attribute !== null && !$this->getDataSet()->hasAttribute($this->attribute);
+        return $this->isDataSetMissing || ($this->attribute !== null && !$this->getDataSet()->hasAttribute($this->attribute));
```
Comment (arogachev) [inline src/ValidationContext.php:273]: > 120 chars.

File: src/ValidationContext.php [hunk 4]
```diff
@@ -127,11 +132,14 @@ public function validate(mixed $data, callable|iterable|object|string|null $rule
 
         $currentDataSet = $this->dataSet;
         $currentAttribute = $this->attribute;
+        $isCurrentDataSetMissing = $this->isDataSetMissing;
 
+        $this->isDataSetMissing = $this->isAttributeMissing();
```
Comment (arogachev) [inline src/ValidationContext.php:137]: `$this->isAttributeMissing`?

Comment (vjik) [reply src/ValidationContext.php:137]: Yes)

File: src/ValidationContext.php (hunk 3 above)
Comment (vjik) [reply src/ValidationContext.php]: done

File: src/ValidationContext.php [hunk 5]
```diff
@@ -127,11 +132,14 @@ public function validate(mixed $data, callable|iterable|object|string|null $rule
 
         $currentDataSet = $this->dataSet;
         $currentAttribute = $this->attribute;
+        $isCurrentDataSetMissing = $this->isDataSetMissing;
```
Comment (arogachev) [inline src/ValidationContext.php:135]: The tests will pass without this line. Maybe remove it or add test for it?

Comment (arogachev) [reply src/ValidationContext.php:135]: The same for `$currentDataSet`, but can be done in a separate PR.

**Review by arogachev — COMMENTED** [review]
See suggestions.

File: src/ValidationContext.php (hunk 4 above)
Comment (arogachev) [reply src/ValidationContext.php:137]: The naming for the property looks illogical for me. Shouldn't it be `$isAttributeMissing`?

Comment (vjik) [reply src/ValidationContext.php:137]: No. Property show missing data set or not.

**Review by arogachev — APPROVED** [review]

File: src/ValidationContext.php (hunk 4 above)
Comment (arogachev) [inline src/ValidationContext.php:137]:
```suggestion
        // The lack of an attribute means that in the context of further validation there is no data set at all.
        $this->isDataSetMissing = $this->isAttributeMissing();
```

**Pull request merged**